### DISCUSSION TITLE
3.1.0

### DISCUSSION THREAD

**Discussion by sfmskywalker** [elsa-workflows/elsa-core#5156] [Design & tech]
## What's Changed
* Merge scope change into main by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4738
* Fix scoping issues by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4753
* New options to control retry logic for transient failures by @NightWuYo in https://github.com/elsa-workflows/elsa-core/pull/4750
* fix EvaluatingCSharp class file name typo by @leeveney in https://github.com/elsa-workflows/elsa-core/pull/4783
* Use NET 8 SDK artifacts output by @lahma in https://github.com/elsa-workflows/elsa-core/pull/4779
* Use correct dialect for SqlServerDbConnectionProvider by @NightWuYo in https://github.com/elsa-workflows/elsa-core/pull/4766
* Fix InvalidOperation Exception for SaveVariablesAsync and DeleteVariablesAsync functions by @NightWuYo in https://github.com/elsa-workflows/elsa-core/pull/4808
* Bump System.Data.SqlClient from 4.8.5 to 4.8.6 in /src/modules/Elsa.Dapper by @dependabot in https://github.com/elsa-workflows/elsa-core/pull/4805
* Allow multiple value processing for Form and QueryString values. by @mohdali in https://github.com/elsa-workflows/elsa-core/pull/4736
* Add NUKE build support by @lahma in https://github.com/elsa-workflows/elsa-core/pull/4815
* Workflow cancellation by @raymonddenhaan in https://github.com/elsa-workflows/elsa-core/pull/4813
* Disabled tests that had an unreliable result by @raymonddenhaan in https://github.com/elsa-workflows/elsa-core/pull/4827
* Bulk Dispatch Workflows Enhancements by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4842
* Drop LinqKit dependency by @Suchiman in https://github.com/elsa-workflows/elsa-core/pull/4853
* Use Directory.Build.props by @lahma in https://github.com/elsa-workflows/elsa-core/pull/4828
* Fix schema for Quartz Postgres Store by @mohdali in https://github.com/elsa-workflows/elsa-core/pull/4864
* Remove Persistence of Completed Activity Execution Contexts by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4869
* Move samples to repository root by @lahma in https://github.com/elsa-workflows/elsa-core/pull/4868
* Refactor logging to exclude activity state by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4874
* Add GZip compression for WorkflowInstance Data field on EF Core by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4876
* Convert to use Central Package Management by @lahma in https://github.com/elsa-workflows/elsa-core/pull/4877
* update sqlserver Extensions support UseCompatibilityLevel cofnig by @Nokecy in https://github.com/elsa-workflows/elsa-core/pull/4844
* Make workflow runtime methods accept nullable options by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4881
* Update InputJsonConverter to deserialize complex value types into their original type by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4878
* issues(4885) by @bbenameur in https://github.com/elsa-workflows/elsa-core/pull/4886
* Fix serialization of delegate expressions by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4888
* issues(4893): try register serializer or skip when already registred … by @bbenameur in https://github.com/elsa-workflows/elsa-core/pull/4894
* Update both published and draft consuming workflows upon child workflow definition update by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4897
* Fix memory leak when using EF Core by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4900
* Update script source path in _Host.cshtml by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4910
* Fix consumption of event payload by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4909
* Fix elsa-workflows/elsa-core#4907 by @platschlena in https://github.com/elsa-workflows/elsa-core/pull/4917
* Fix WorkflowExecutionContext bug by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4919
* Update AppAny package version and fix MySql and SqlServer Quartz persistence by @mohdali in https://github.com/elsa-workflows/elsa-core/pull/4921
* Fix serialization of Boolean activity input by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4922
* issues(4918): fix resolve scoped service for Elsa.MongoDb.Modules by @bbenameur in https://github.com/elsa-workflows/elsa-core/pull/4924
* Improve Elsa workflow expression serialization by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4930
* Implement dispatch channels by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4949
* Fixed setting updated and finished date on workflow instance by @raymonddenhaan in https://github.com/elsa-workflows/elsa-core/pull/4943
* Updated client to include inputs as part of workflow instance by @MariusVuscanNx in https://github.com/elsa-workflows/elsa-core/pull/4955
* Set the Timer Interval default value to match DefaultValue by @mohdali in https://github.com/elsa-workflows/elsa-core/pull/4926
* feature: Add created date time to filter by @bbenameur in https://github.com/elsa-workflows/elsa-core/pull/4932
* Removed references to packages that are not used by @Inx51 in https://github.com/elsa-workflows/elsa-core/pull/4952
* Add PR workflow by @lahma in https://github.com/elsa-workflows/elsa-core/pull/4956
* Enhanced Workflow Instance Filters by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4957
* #4963 - elastic store workflowinstance search fix by @ayalros in https://github.com/elsa-workflows/elsa-core/pull/4964
* Introduced messaging to running instances by @raymonddenhaan in https://github.com/elsa-workflows/elsa-core/pull/4941
* Implement distributed lock for registry population by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4983
* Added workflow instance and bulk cancellation to the UI by @raymonddenhaan in https://github.com/elsa-workflows/elsa-core/pull/4965
* Example: added MassTransit triggered workflow example-project (#4968) by @ckeller81 in https://github.com/elsa-workflows/elsa-core/pull/4981
* Fix Dapper Provider by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4998
* Update BulkUpsertAsync extension method with EFCore.BulkExtensions package by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/5005
* Respond with 500 in case of Incidents by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/5004
* Fix Security Bug: No permissions applied to version endpoints by @IsaacHayes1995 in https://github.com/elsa-workflows/elsa-core/pull/5008
* Fix incorrect comment information in EFCoreWorkflowInboxMessageStore by @faith0831 in https://github.com/elsa-workflows/elsa-core/pull/4979
* Added logging to pinpoint issue #5033 by @raymonddenhaan in https://github.com/elsa-workflows/elsa-core/pull/5034
* Add SearchTerm to Workflow Definitions List API client by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/5035
* fix missing metadata when persisting bookmark by @carbonhe in https://github.com/elsa-workflows/elsa-core/pull/5026
* Update VersionOptions.LatestAndPublished to set correct property by @pramsperger in https://github.com/elsa-workflows/elsa-core/pull/5036
* Improved search to not require exact match by @MariusVuscanNx in https://github.com/elsa-workflows/elsa-core/pull/5037
* Added Properties field to ActivityExecutionRecord by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/5039
* Reverted bulk implementation by @raymonddenhaan in https://github.com/elsa-workflows/elsa-core/pull/5047
* Add Unicode encoding support to serializers by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/5049
* Added check preventing duplicate workflow to update by @raymonddenhaan in https://github.com/elsa-workflows/elsa-core/pull/5060
* Enhance: Add isStart property to determine if current node is a start… by @zergmk2 in https://github.com/elsa-workflows/elsa-core/pull/5040
* Add Unicode string serialization tests by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/5064
* Added missing text parser for HTTP responses by @raymonddenhaan in https://github.com/elsa-workflows/elsa-core/pull/5066
* Prevented overwriting of existing workflows by @raymonddenhaan in https://github.com/elsa-workflows/elsa-core/pull/5067
* Refactor workflow instance listing and enum parsing by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/5069
* Alterations with Queries by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/5032
* Made loading workflows more robust by @raymonddenhaan in https://github.com/elsa-workflows/elsa-core/pull/5074
* Update MongoDb collection service retrieval by @daveelbro in https://github.com/elsa-workflows/elsa-core/pull/5083
* Update VersionOptions.cs by @pietervoorschot in https://github.com/elsa-workflows/elsa-core/pull/5053
* Add configuration value accessor in JintJavaScriptEvaluator by @nuvotex-tk in https://github.com/elsa-workflows/elsa-core/pull/5024
* Fix ID assignment logic in Scheduler by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/5094
* Test workflows housekeeping by @MariusVuscanNx in https://github.com/elsa-workflows/elsa-core/pull/5092
* Add alteration notifications and management by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/5097
* Add log message sanitization for PostgreSQL by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/5104
* Add DryRun endpoint and workflow instance finder by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/5110
* Add logging to ScheduledCronTask constructor and setup by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/5122
* Implemented distributed cache refresh for workflow definitions by @raymonddenhaan in https://github.com/elsa-workflows/elsa-core/pull/5095
* Feature/cancel dispatched workflows by @raymonddenhaan in https://github.com/elsa-workflows/elsa-core/pull/5136
* Prevented creation of default queues for temporary consumers by @raymonddenhaan in https://github.com/elsa-workflows/elsa-core/pull/5148
* Fix concurrency issues with distributed messages by @raymonddenhaan in https://github.com/elsa-workflows/elsa-core/pull/5149
* #5107 Add framework reference & package removal by @thompson-tomo in https://github.com/elsa-workflows/elsa-core/pull/5128
* #5106 Cleanup packages to leverage framework by @thompson-tomo in https://github.com/elsa-workflows/elsa-core/pull/5129
* Update FlowNode status terminology in BulkDispatchWorkflows by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/5150
* Make JetBrains.Annotations Private by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/5151
* Fix Alteration Runner by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/5142
* Fix flowchart completion after faulted child activity cancellation by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/5145
* Fixed per instance distributed messaging by @raymonddenhaan in https://github.com/elsa-workflows/elsa-core/pull/5154
* Regenerate V3.1 EF Core migrations with custom runtime schema support by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/5155

## New Contributors
* @leeveney made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/4783
* @lahma made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/4779
* @raymonddenhaan made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/4813
* @Suchiman made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/4853
* @bbenameur made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/4886
* @platschlena made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/4917
* @Inx51 made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/4952
* @ayalros made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/4964
* @IsaacHayes1995 made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/5008
* @faith0831 made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/4979
* @carbonhe made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/5026
* @pramsperger made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/5036
* @daveelbro made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/5083
* @pietervoorschot made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/5053
* @nuvotex-tk made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/5024
* @thompson-tomo made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/5128

**Full Changelog**: https://github.com/elsa-workflows/elsa-core/compare/3.0.6...3.1.0

<hr /><em>This discussion was created from the release <a href='https://github.com/elsa-workflows/elsa-core/releases/tag/3.1.0'>3.1.0</a>.</em>

**Comment by glime-ai** [comment]
The Elsa Workflows project has released version 3.1.0 with numerous changes and improvements. Here are some of the highlights:

- Scope changes and scoping issues have been addressed.
- New options to control retry logic for transient failures have been added.
- Various fixes, including a typo in the EvaluatingCSharp class file name and issues with SqlServerDbConnectionProvider.
- Improvements to the handling of Form and QueryString values, allowing for multiple value processing.
- Addition of NUKE build support and workflow cancellation features.
- Enhancements to the Bulk Dispatch Workflows.
- Dependency updates, such as bumping System.Data.SqlClient from 4.8.5 to 4.8.6.
- Architectural changes, like dropping the LinqKit dependency and using Directory.Build.props.
- Fixes for memory leaks when using EF Core and improvements to the WorkflowExecutionContext.
- Introduction of messaging to running instances and distributed lock for registry population.
- UI updates, including workflow instance and bulk cancellation features.
- Various other bug fixes, performance improvements, and code refactoring.

New contributors have also joined the project, making their first contributions in this release cycle.

For a full list of changes, you can view the complete changelog on the GitHub release page: https://github.com/elsa-workflows/elsa-core/compare/3.0.6...3.1.0
